sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java

```diff
@@ -0,0 +1,747 @@
+ * </ul>
+ *
+ * <h4>Query the resulting {@link PCollection} containing the sketch.</h4>
+ *
+ * <pre><code>
+ * int k = 100;
+ * int capacity = 100000
+ * {@literal PCollection<StreamSummarySketch<MyClass>>} input = ...;
+ * {@literal PCollection<KV<MyClass, Long>>} output = input
+ *     .apply(MostFrequent.MostFrequentFn.NonSerializableElements
+ *        .create(new MyClassCoder())
+ *        .withCapacity(capacity))
+ *     .apply(ParDo.of(RetrieveTopK.globallyNonSerializable()));
+ * </code></pre>
+ *
+ * <b>Warning: this class is experimental.</b> <br>
+ * Its API is subject to change in future versions of Beam.
+ */
+@Experimental
+public final class MostFrequent {
+
+  /**
+   * Computes and returns the top K most frequent elements in the input {@link PCollection}.
+   *
+   * @param <InputT>    the type of the elements in the input {@link PCollection}
+   */
+  public static <InputT> GlobalSummary<InputT> globally() {
+    return GlobalSummary.<InputT>builder().build();
+  }
+
+  /**
+   * Computes and returns the top K most frequent elements in the input
+   * {@link PCollection} of {@link KV}s.
+   *
+   * @param <K>         the type of the keys mapping the elements
+   * @param <V>         the type of the elements being combined per key
+   */
+  public static <K, V> PerKeySummary<K, V> perKey() {
+    return PerKeySummary.<K, V>builder().build();
+  }
+
+  /** Implementation of {@link #globally()}. */
+  @AutoValue
+  public abstract static class GlobalSummary<InputT>
+          extends PTransform<PCollection<InputT>, PCollection<KV<InputT, Long>>> {
+
+    abstract int capacity();
+    abstract int topK();
+    abstract Builder<InputT> toBuilder();
+
+    static <InputT> Builder<InputT> builder() {
+      return new AutoValue_MostFrequent_GlobalSummary.Builder<InputT>()
+              .setCapacity(10000)
+              .setTopK(100);
+    }
+
+    @AutoValue.Builder
+    abstract static class Builder<InputT> {
+      abstract Builder<InputT> setCapacity(int c);
+      abstract Builder<InputT> setTopK(int k);
+
+      abstract GlobalSummary<InputT> build();
+    }
+
+    /**
+     * Sets the number {@code k} of most frequent element to retrieve.
+     *
+     * @param k the top k most frequent elements to return
+     */
+    public GlobalSummary<InputT> topKElements(int k) {
+      return toBuilder().setTopK(k).build();
+    }
+
+    /**
+     * Sets the capacity {@code c}.
+     *
+     * @param c the capacity of the summary
+     */
+    public GlobalSummary<InputT> withCapacity(int c) {
+      return toBuilder().setCapacity(c).build();
+    }
+
+    @Override
+    public PCollection<KV<InputT, Long>> expand(PCollection<InputT> input) {
+      final Coder<InputT> inputCoder = input.getCoder();
+      Class clazz = inputCoder.getEncodedTypeDescriptor().getRawType();
+      PCollection<KV<InputT, Long>> result;
+
+      if (Serializable.class.isAssignableFrom(clazz)) {
```

Comment:
Why is this needed? And why do we ignore the input coder if the class is serializable?

Reply:
I wanted to distinguish the cases where elements are serializable from cases where elements are not serializable, because in the latter case I'll have to wrap elements in a serializable wrapper.
This is due to the serialization of StreamSummary which serializes the tracked elements using the default method "outSream.writeObject()"
I cannot change the way the StreamSummary is serialized because of private some members.

In the first case elements don't need to be wrapped therefore it would be a loss of time to wrap them, that's why I made a distinction. 

Another solution would be to encode elements into bytes before offering them to the structure (like we did for Hyperloglog for instance) but I think this will be slower

Reply:
My objection is to this:

> I cannot change the way the StreamSummary is serialized because of some private members.

I believe you don't have to serialize StreamSummary directly. Instead, you can develop a Coder for it that works like this:

encode(): call topK(Integer.MAX_VALUE), obtain a list of element/count pairs, serialize those (using the user's provided input coder)
decode(): get the element/count pairs, insert each of them into a new empty StreamSummary, return that.

Reply:
You're right, I also thought about this but don't you think this will drastically slow down the Combine processing ?